low-level/sensors/impl/centos_7/service_monitor.py

```diff
@@ -216,9 +227,8 @@ def properties_changed_handler(self, interface, changed_properties,
 
         if self.state != self.previous_state:
             if self.state == "active":
-                if self._service_state is FailedState:
-                    Service.alerts.put(
-                        ServiceMonitor.get_alert(self, ResolvedAlert))
+                if self._service_state is FailedState and self.name not in Service.active_services:
+                    Service.active_services.add(self.name)
```

Comment:
why service added in active list, when its state is failed?
Also raising alert is avoided here, Line 221 ?

Reply:
`why service added in active list, when its state is failed?`

We are adding service in active_services set if current state of service is 'active'. Line-229.

`Also raising alert is avoided here, Line 221 ?`
this has been moved to **check_active_services** function. Line-464

Reply:
>We are adding service in active_services set if current state of service is 'active'. Line-229.
I dont understand this still, line 229 must have changed with code changes, no code there now.

Reply:
Okay let me explain with example.
If service detected Failed/Inactive for threshold time, service state changed to FailedState (self._service_state=FailedState). 
Now when in next iteration service state change detected, 
here in properties_changed_handler function we fetch current state,substate pid of service.
then check if current state not equals to prev_state, and if current service state is active then check
 self.service_state = FailedState and service name not active service list, then only we add it into active_service set.
 and in next line chnged service state to self.service_state = ActiveState. 
 
```
 if self.state != self.previous_state:
            if self.state == "active":
                if self._service_state is FailedState and self.name not in Service.active_services:
                    Service.active_services.add(self.name)
                self.new_service_state(ActiveState)
```

Reply:
Understand now, just one comment. Better to do
>self.new_service_state(ActiveState)

first, before
> self.new_service_state(ActiveState)

Reply:
`Service.active_services.add(self.name)`

Can't you  add this into ActiveService class's enter method ?

Reply:
Service get added into list only if there is previous fault_alert detected for service and current state is active, only for such service we raise fault_resolved alert.
If I move this to ActiveService class's enter method then all active services gets added into this list and we will get fault_resolved alert for all this services even if there is no fault alert detected.